src/KubernetesClient/Informers/ResourceObservableExtensions.cs

```diff
@@ -0,0 +1,299 @@
+            this IObservable<ResourceEvent<TResource>> source,
+            ICache<TKey, TResource> cache,
+            Func<TResource, TKey> keySelector,
+            Func<TResource, TResource, TResource> mapper = null)
+        {
+            if (mapper == null)
+                mapper = (oldResource, newResource) => newResource;
+
+            return Observable.Defer(() =>
+            {
+                long msgNum = 0;
+
+                return source
+                    .Do(_ => msgNum++)
+                    .WithResets(events =>
+                    {
+                        var reset = events
+                            .Select(x => x.Value)
+                            .Where(x => x != null)
+                            .ToDictionary(keySelector, x => x);
+
+                        cache.Reset(reset);
+                        cache.Version += events.Count;
+                        return events;
+                    }, notification =>
+                    {
+                        if (notification.EventFlags.HasFlag(EventTypeFlags.Reset))
+                        {
+                        }
+                        else if (!notification.EventFlags.HasFlag(EventTypeFlags.Delete))
+                        {
+                            cache.Version++;
+                            if (notification.Value != null)
+                            {
+                                var key = keySelector(notification.Value);
+                                if (cache.TryGetValue(key, out var existing))
+                                    notification = new ResourceEvent<TResource>(notification.EventFlags, mapper(existing, notification.Value), existing);
+                                cache[keySelector(notification.Value)] = notification.Value;
+                            }
+                        }
+                        else
+                        {
+                            cache.Remove(keySelector(notification.OldValue));
+                        }
+
+                        return new CacheSynchronized<ResourceEvent<TResource>>(msgNum, cache.Version, notification);
+                    });
+            });
+        }
+
+
+        public static IObservable<ResourceEvent<TResource>> ComputeMissedEventsBetweenResets<TKey, TResource>(this IObservable<ResourceEvent<TResource>> source, Func<TResource, TKey> keySelector, IEqualityComparer<TResource> comparer)
+        {
+            return Observable.Create<ResourceEvent<TResource>>(observer =>
+            {
+                Dictionary<TKey, TResource> cacheSnapshot;
+                var cache = new SimpleCache<TKey, TResource>();
+                var cacheSynchronized = false;
+                return source
+                    .WithResets(resetBuffer =>
+                    {
+                        if (!cacheSynchronized) return resetBuffer;
+
+                        cacheSnapshot = new Dictionary<TKey, TResource>(cache);
```

Comment:
Consider moving `cacheSnapshot` declaration here; otherwise, it's unnecessarily rooted for the lifetime of the subscription.

Reply:
Can this be `cache.Snapshot()` instead?

Reply:
Good catch. This code was written before I introduced Snapshot as part of ICache.